ADAL/src/request/ADAuthenticationRequest+AcquireToken.m

```diff
@@ -431,84 +439,139 @@ - (void)requestTokenImpl:(ADAuthenticationCallback)completionBlock
-                                   completionHandler:completionBlock];
-                 return;
-             }
-             else
+        if ([response isKindOfClass:MSIDWebMSAuthResponse.class])
+        {
+            [event setAPIStatus:@"try to prompt to install broker"];
+            [[MSIDTelemetry sharedInstance] stopEvent:_requestParams.telemetryRequestId event:event];
+            
+            ADAuthenticationError *error = nil;
+            NSURL* brokerRequestURL = [self composeBrokerRequest:&error];
+            if (!brokerRequestURL)
+            {
+                ADAuthenticationResult *result = [ADAuthenticationResult resultFromError:error correlationId:_requestParams.correlationId];
+                [result setCloudAuthority:_cloudAuthority];
+                completionBlock(result);
+                return;
+            }
+            
+            [ADBrokerHelper promptBrokerInstall:[NSURL URLWithString:((MSIDWebMSAuthResponse *)response).appInstallLink]
+                                  brokerRequest:brokerRequestURL
+                              completionHandler:completionBlock];
+            return;
+        }
 #endif
-             {
-                 [event setAPIStatus:MSID_TELEMETRY_VALUE_SUCCEEDED];
-                 [[MSIDTelemetry sharedInstance] stopEvent:_requestParams.telemetryRequestId event:event];
-                 
-                 [[MSIDTelemetry sharedInstance] startEvent:_requestParams.telemetryRequestId eventName:MSID_TELEMETRY_EVENT_TOKEN_GRANT];
-                 [self requestTokenByCode:code
-                          completionBlock:^(MSIDTokenResponse *response, ADAuthenticationError *error)
-                  {
-                      ADAuthenticationResult *result = [ADResponseCacheHandler processAndCacheResponse:response
-                                                                                      fromRefreshToken:nil
-                                                                                                 cache:self.tokenCache
-                                                                                                params:_requestParams];
-                      
-                      [result setCloudAuthority:_cloudAuthority];
-                      
-                      ADTelemetryAPIEvent *event = [[ADTelemetryAPIEvent alloc] initWithName:MSID_TELEMETRY_EVENT_TOKEN_GRANT
-                                                                                     context:_requestParams];
-                      [event setGrantType:MSID_TELEMETRY_VALUE_BY_CODE];
-                      [event setResultStatus:[result status]];
-                      [[MSIDTelemetry sharedInstance] stopEvent:_requestParams.telemetryRequestId event:event];
-                      
-                      completionBlock(result);
-                  }];
-             }
-         }
-     }];
+        
+        if ([response isKindOfClass:MSIDWebOpenBrowserResponse.class])
```

Comment:
Since we're changing this code, can we separate it into sub-methods? It's a huge completion block right now and it's a bit difficult to follow :) e.g. openBrowserURL:(NSURL*)url withCompletionHandler:... could check if we're running in extension and fail if necessary etc...

Reply:
Refactored to sub methods for handling the actual responses per class type.